File: scripts/plans.mjs
import { execSync } from "child_process";
import { createWriteStream, mkdirSync } from "fs";
import { dirname, join } from "path";
import { fileURLToPath } from "url";

const OUTPUT_DIR = join(dirname(fileURLToPath(import.meta.url)), "../public/api");
mkdirSync(OUTPUT_DIR, { recursive: true });
const OUTPUT = join(OUTPUT_DIR, "plans.csv");

/** @returns {unknown[]} */
function fetchPlansForYear(/** @type {number} */ year) {
  const url = `https://api.hpc.tools/v2/public/plan?year=${year}&limit=500`;
  // Node's default "node" User-Agent is blocked by the HPC API bot filter;
  // use curl which is on the API's allowlist.
  const stdout = execSync(`curl -sf "${url}"`, { timeout: 30_000, encoding: "utf8" });
  return JSON.parse(stdout).data;
}

const FIRST_YEAR = 2000;

function fetchPlans() {
  const currentYear = new Date().getFullYear();
  const years = Array.from({ length: currentYear - FIRST_YEAR + 1 }, (_, i) => currentYear - i);
  /** @type {{ iso3: string; year: number; type: string; id: number }[]} */
  const allRows = [];
  for (const year of years) {
    const plans = fetchPlansForYear(year);
    console.log(`${plans.length} plans fetched for ${year}`);
    allRows.push(...buildRows(plans, year));
  }
  return allRows;
}

/** @param {unknown} plan */
function getPlanType(plan) {
  const p = /** @type {any} */ (plan);
  for (const cat of p.categories ?? []) {
    if (cat.group === "planType") return cat.code ?? "";
  }
  return "";
}

/**
 * @param {unknown[]} plans
 * @param {number} year
 */
function buildRows(plans, year) {
  /** @type {{ iso3: string; year: number; type: string; id: number }[]} */
  const rows = [];
  for (const plan of plans) {
    const p = /** @type {any} */ (plan);
    const type = getPlanType(p);
    for (const loc of p.locations ?? []) {
      const iso3 = loc.iso3 ?? "";
      if (iso3) rows.push({ iso3, year, type, id: p.id });
    }
  }
  return rows.sort(
    (a, b) =>
      a.iso3.localeCompare(b.iso3) ||
      a.type.localeCompare(b.type) ||
      String(a.id).localeCompare(String(b.id)),
  );
}

/**
 * @param {{ iso3: string; year: number; type: string; id: number }[]} rows
 * @param {string} path
 */
function writeCsv(rows, path) {
  const fields = /** @type {const} */ (["iso3", "year", "type", "id"]);
  const out = createWriteStream(path);
  out.write(fields.join(",") + "\n");
  for (const row of rows) {
    out.write(fields.map((f) => row[f]).join(",") + "\n");
  }
  out.end();
}

const rows = fetchPlans();
writeCsv(rows, OUTPUT);
console.log(`${rows.length} rows written to ${OUTPUT}`);

File: scripts/plans.py
# /// script
# requires-python = ">=3.12"
# dependencies = ["requests"]
# ///

import csv
from datetime import datetime
from pathlib import Path

import requests

OUTPUT_DIR = Path(__file__).parent.parent / "public" / "api"
OUTPUT_DIR.mkdir(parents=True, exist_ok=True)
OUTPUT = OUTPUT_DIR / "plans.csv"

FIRST_YEAR = 2000


def fetch_plans_for_year(year: int) -> list:
    url = f"https://api.hpc.tools/v2/public/plan?year={year}&limit=500"
    res = requests.get(url, timeout=30)
    res.raise_for_status()
    return res.json().get("data", [])


def get_plan_type(plan: dict) -> str:
    for cat in plan.get("categories", []):
        if cat.get("group") == "planType":
            return cat.get("code", "")
    return ""


def build_rows(plans: list, year: int) -> list[dict]:
    rows = []
    for plan in plans:
        plan_type = get_plan_type(plan)
        for loc in plan.get("locations", []):
            iso3 = loc.get("iso3", "")
            if iso3:
                rows.append({"iso3": iso3, "year": year, "type": plan_type, "id": plan["id"]})
    return sorted(rows, key=lambda r: (r["iso3"], r["type"], str(r["id"])))


current_year = datetime.now().year
years = list(range(current_year, FIRST_YEAR - 1, -1))

all_rows: list[dict] = []
for year in years:
    plans = fetch_plans_for_year(year)
    print(f"{len(plans)} plans fetched for {year}")
    all_rows.extend(build_rows(plans, year))

with open(OUTPUT, "w", newline="", encoding="utf-8-sig") as f:
    writer = csv.DictWriter(f, fieldnames=["iso3", "year", "type", "id"])
    writer.writeheader()
    writer.writerows(all_rows)

print(f"{len(all_rows)} rows written to {OUTPUT}")
